pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+		s.db.WriteTxn(s.svcs, append(extraTables, s.bes)...),
+	}
+}
+
+func (s *Services) UpsertService(txn ServiceWriteTxn, params *ServiceParams) (created bool, err error) {
+	var svc Service
+
+	existing, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(params.Frontend))
+	if found {
+		// Do not allow services with same address but different names to override each other.
+		if !existing.Name.Equal(params.Name) {
+			return false, ErrServiceConflict
+		}
+
+		// Do not allow overriding a service that has been created from
+		// another source. If the service is 'Unspec'
+		if existing.Source != params.Source {
+			return false, ErrServiceSourceMismatch
+		}
+
+		// Merge the existing fields.
+		svc = *existing
+	}
+	created = !found
+
+	svc.ServiceParams = params
+	svc.Status = reconciler.StatusPending()
+
+	// Run hooks
+	for _, hook := range s.hooks {
+		hook(txn, &svc)
+	}
+
+	_, _, err = s.svcs.Insert(txn, &svc)
+	return
+}
+
+func (s *Services) DeleteService(txn ServiceWriteTxn, name ServiceName, addr L3n4Addr) (old *Service, err error) {
+	old, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(addr))
+	if !found {
+		return nil, nil
+	}
+	if !old.Name.Equal(name) {
+		return nil, ErrServiceConflict
+	}
+	err = s.deleteService(txn, old)
+	return
+}
+
+func (s *Services) DeleteServicesByName(txn ServiceWriteTxn, name ServiceName, source source.Source) error {
+	svc, _, found := s.svcs.First(txn, ServiceNameIndex.Query(name))
+	if !found {
+		return statedb.ErrObjectNotFound
+	}
+	if svc.Source != source {
+		return ErrServiceSourceMismatch
+	}
+
+	return s.deleteService(txn, svc)
+}
+
+func (s *Services) deleteService(txn ServiceWriteTxn, svc *Service) error {
+	// Release references to the backends
+	iter, _ := s.bes.Get(txn, BackendServiceIndex.Query(svc.Name))
+	for be, _, ok := iter.Next(); ok; be, _, ok = iter.Next() {
+		be, orphan := be.removeRef(svc.Name)
+		if orphan {
+			if _, _, err := s.bes.Delete(txn, be); err != nil {
+				return err
+			}
+		} else {
+			if _, _, err := s.bes.Insert(txn, be); err != nil {
+				return err
+			}
+		}
+	}
+
+	svc.Status = reconciler.StatusPendingDelete()
+	_, _, err := s.svcs.Insert(txn, svc)
+	return err
+}
+
+func (s *Services) DeleteServicesBySource(txn ServiceWriteTxn, source source.Source) error {
+	// Iterating over all as this is a rare operation and it would be costly
+	// to always index by source.
+	iter, _ := s.svcs.All(txn)
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+		if svc.Source == source {
+			if err := s.deleteService(txn, svc); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Services) UpsertBackends(txn ServiceWriteTxn, serviceName ServiceName, source source.Source, bes ...*Backend) error {
+	if err := s.updateBackends(txn, serviceName, source, bes); err != nil {
+		return err
+	}
+
+	iter, _ := s.svcs.Get(txn, ServiceNameIndex.Query(serviceName))
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
```

Comment:
No because `Service` in this context really only describes a single frontend. There can be many of them sharing the same name, but having different type, IP or port. Same as how it works currently with `loadbalancer.SVC` and ServiceManager. A case could be made to instead call these things frontends...

Reply:
I've now renamed `Service` to `Frontend`.